internal/fwserver/server_upgraderesourceidentity.go

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package fwserver
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-framework/internal/fwschema"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/internal/logging"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+)
+
+// UpgradeResourceIdentityRequest is the framework server request for the
+// UpgradeResourceIdentity RPC.
+type UpgradeResourceIdentityRequest struct {
+	Resource       resource.Resource
+	IdentitySchema fwschema.Schema
+	// TypeName is the type of resource that Terraform needs to upgrade the
+	// identity state for.
+	TypeName string
+
+	// Version is the version of the identity state the resource currently has.
+	Version int64
+
+	// Using the tfprotov6 type here was a pragmatic effort decision around when
+	// the framework introduced compatibility promises. This type was chosen as
+	// it was readily available and trivial to convert between tfprotov5.
+	//
+	// Using a terraform-plugin-go type is not ideal for the framework as almost
+	// all terraform-plugin-go types have framework abstractions, but if there
+	// is ever a time where it makes sense to re-evaluate this decision, such as
+	// a major version bump, it could be changed then.
+	// Reference: https://github.com/hashicorp/terraform-plugin-framework/issues/340
+	RawState *tfprotov6.RawState
+}
+
+// UpgradeResourceIdentityResponse is the framework server response for the
+// UpgradeResourceIdentity RPC.
+type UpgradeResourceIdentityResponse struct {
+	UpgradedIdentity *tfsdk.ResourceIdentity
+	Diagnostics      diag.Diagnostics
+}
+
+// UpgradeResourceIdentity implements the framework server UpgradeResourceIdentity RPC.
+func (s *Server) UpgradeResourceIdentity(ctx context.Context, req *UpgradeResourceIdentityRequest, resp *UpgradeResourceIdentityResponse) {
+	if req == nil {
+		return
+	}
+
+	// No UpgradedIdentity to return. This could return an error diagnostic about
+	// the odd scenario, but seems best to allow Terraform CLI to handle the
+	// situation itself in case it might be expected behavior.
+	if req.RawState == nil {
+		return
+	}
+
+	// Define options to be used when unmarshalling raw state.
+	// IgnoreUndefinedAttributes will silently skip over fields in the JSON
+	// that do not have a matching entry in the schema.
+	unmarshalOpts := tfprotov6.UnmarshalOpts{
+		ValueFromJSONOpts: tftypes.ValueFromJSONOpts{
+			IgnoreUndefinedAttributes: true,
+		},
+	}
+
+	// TODO: Maybe throw error if the schemas are the same, it is a bug in core
```

Comment:
I think this sounds like a good idea to do 🙂 

All of the commented code can go as well 👍🏻